File: 283.move-zeroes.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * @lc app=leetcode id=283 lang=cpp
 *
 * [283] Move Zeroes
 */

// @lc code=start
class Solution {
   public:
    void moveZeroes(vector<int>& nums) {
        int n = nums.size();
        if (n == 1) return;
        if (n == 2 && nums[0] == 0) {
            swap(nums[0], nums[1]);
            return;
        }
        int replaceIndex = -1, zeroCnt = 0;
        for (int i = 0; i < n; i++) {
            if (nums[i] == 0) {
                if (replaceIndex == -1) replaceIndex = i;
                zeroCnt++;
            } else {
                if (replaceIndex != -1) {
                    nums[replaceIndex++] = nums[i];
                }
            }
        }
        int i = n - 1;
        while (zeroCnt--) {
            nums[i--] = 0;
        }
    }
};
// @lc code=end

File: java/283.move-zeroes.java
/*
 * @lc app=leetcode id=283 lang=java
 *
 * [283] Move Zeroes
 */

// @lc code=start
class Solution {
    public void moveZeroes(int[] nums) {
        int i = 0, j = 1;
        // 1 2 0 3 4
        while (i < nums.length && nums[i] != 0) {
            i++;
        }
        j = i + 1;
        while (j < nums.length) {
            if (nums[j] != 0)
                nums[i++] = nums[j++];
            else
                j++;
        }
        while (i < nums.length)
            nums[i++] = 0;
    }
}
// @lc code=end

